server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+    if ((tabletEndRow == null
+        || !tabletEndRow.equals(deleteRange.toDataRange().getEndKey().getRow()))
+        && deleteRange.toDataRange().getEndKey() != null
+        && tabletRange.contains(deleteRange.toDataRange().getEndKey())) {
+      Manager.log.trace(
+          "Fencing tablet at end of deletion range with end row {}, End deletion row {}",
+          tabletEndRow, deleteRange.toDataRange().getEndKey().getRow());
+      ranges.add(new Range(deleteRange.toDataRange().getEndKey().getRow(), tabletEndRow));
+    }
+
+    return ranges;
+  }
+
+  private void updateMetadataRecordsForDelete(MergeInfo info) throws AccumuloException {
+    final KeyExtent range = info.getExtent();
+
+    String targetSystemTable = MetadataTable.NAME;
+    if (range.isMeta()) {
+      targetSystemTable = RootTable.NAME;
+    }
+
+    final AccumuloClient client = manager.getContext();
+
+    try (BatchWriter bw = client.createBatchWriter(targetSystemTable)) {
+      final Text startRow = range.prevEndRow();
+      final Text lastRow = range.endRow() != null
+          ? new Key(range.endRow()).followingKey(PartialKey.ROW).getRow() : null;
+
+      // Find the tablets that overlap the start and end of the deletion range
+      final List<TabletMetadata> tabletMetadatas = new ArrayList<>();
+      loadTabletMetadata(range.tableId(), startRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tabletMetadatas::add);
+      loadTabletMetadata(range.tableId(), lastRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tabletMetadatas::add);
+
+      for (TabletMetadata tabletMetadata : tabletMetadatas) {
+        final KeyExtent keyExtent = tabletMetadata.getExtent();
+
+        // Check if this tablet needs to have its files fenced for the deletion
+        if (needsFencingForDeletion(info, keyExtent)) {
+          Manager.log.debug("Found overlapping keyExtent {} for delete, fencing files.", keyExtent);
+
+          // Create the ranges for fencing the files, this takes the place of
+          // chop compactions and splits
+          final List<Range> ranges = createRangesForDeletion(tabletMetadata, range);
+
+          // Go through and fence each of the files that are part of the tablet
+          for (Entry<StoredTabletFile,DataFileValue> entry : tabletMetadata.getFilesMap()
+              .entrySet()) {
+            StoredTabletFile existing = entry.getKey();
+            Value value = entry.getValue().encodeAsValue();
+
+            Mutation m = new Mutation(keyExtent.toMetaRow());
+
+            // Go through each range that was created and modify the metadata for the file
+            // The end row should be inclusive for the current tablet and the previous end row
+            // should be exclusive for the start row.
+            for (Range fenced : ranges) {
+              // Clip range with the tablet range if the range already exists
+              fenced = existing.hasRange() ? existing.getRange().clip(fenced) : fenced;
```

Comment:
The ranges could be disjoint which would throw an exception,  thinking if they are disjoint then just drop the file w/o adding anything?  It would be cleaner to put the rest of the code in an else instead of doing a continue like the code below, was just trying to make the suggestion short.

```suggestion
              fenced = existing.hasRange() ? existing.getRange().clip(fenced, true) : fenced;

              if(fenced == null) {
                m.putDelete(DataFileColumnFamily.NAME, existing.getMetadataText());
                continue;
              }
```

Reply:
I actually just ran into this exact situation as I was testing deleting with tablets that already have ranged files so I will be making sure to include this with my next testing updates.

Reply:
This has been fixed and includes new tests in https://github.com/apache/accumulo/pull/3728/commits/0e6c5d70bc1fad6b234206e103d1851d52e752df

If a file is disjoint for all the new range(s) then it will be dropped. There maybe 2 ranges created if the deletion range is contained within 1 tablet so we need to check both. the ITs include tests for deleting existing ranged files